File: Exp_2/HC.java
import java.util.Scanner;

public class HC {
    private final String alphaList;
    private final int m;

    public HC(int m, String alphaList) {
        this.alphaList = alphaList;
        this.m = m;
    }

    private int findGCD(int a, int b) {
        if (a < b) {
            int temp = a;
            a = b;
            b = temp;
        }
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    private int extendedEuclids(int a, int b, int m) {
        int t1 = 0, t2 = 1;
        if (a < b) {
            int temp = a;
            a = b;
            b = temp;
        }
        while (b != 0) {
            int q = a / b;
            int temp = b;
            b = a % b;
            a = temp;
            int t = t1 - q * t2;
            t1 = t2;
            t2 = t;
        }
        if (t1 < 0) {
            return t1 + m;
        } else {
            return t1;
        }
    }

    private int[][] convertStrMat(String word, int syllable) {
        while (word.length() % syllable != 0) {
            word += 'X';
        }
        int rows = syllable;
        int cols = word.length() / syllable;
        int[][] result = new int[rows][cols];
        int k = 0;
        for (int i = 0; i < cols; i++) {
            for (int j = 0; j < rows; j++) {
                result[j][i] = alphaList.indexOf(word.charAt(k++));
            }
        }
        return result;
    }

    private String convertMatStr(int[][] mat) {
        StringBuilder text = new StringBuilder();
        for (int i = 0; i < mat[0].length; i++) {
            for (int j = 0; j < mat.length; j++) {
                text.append(alphaList.charAt(mat[j][i] % m));
            }
        }
        return text.toString();
    }

    private int[][] findMatMul(int[][] matA, int[][] matB) {
        int rowsA = matA.length;
        int colsA = matA[0].length;
        int colsB = matB[0].length;
        int[][] result = new int[rowsA][colsB];
        for (int i = 0; i < rowsA; i++) {
            for (int j = 0; j < colsB; j++) {
                for (int k = 0; k < colsA; k++) {
                    result[i][j] += matA[i][k] * matB[k][j];
                }
                result[i][j] %= m;
            }
        }
        return result;
    }

    private int[][] removeCopyMatrix(int[][] mat, int i, int j) {
        int size = mat.length - 1;
        int[][] result = new int[size][size];
        for (int row = 0, newRow = 0; row < mat.length; row++) {
            if (row == i) continue;
            for (int col = 0, newCol = 0; col < mat[0].length; col++) {
                if (col == j) continue;
                result[newRow][newCol++] = mat[row][col];
            }
            newRow++;
        }
        return result;
    }

    private boolean checkSquareMatrix(int[][] mat) {
        for (int[] row : mat) {
            if (row.length != mat.length) return false;
        }
        return true;
    }

    private int findDeterminant(int[][] mat) {
        if (!checkSquareMatrix(mat)) return -1;
        if (mat.length == 2) {
            return (mat[0][0] * mat[1][1] - mat[0][1] * mat[1][0]) % m;
        }
        int det = 0;
        for (int c = 0; c < mat.length; c++) {
            det += ((c % 2 == 0 ? 1 : -1) * mat[0][c] * findDeterminant(removeCopyMatrix(mat, 0, c))) % m;
        }
        return det < 0 ? det + m : det;
    }

    private int[][] findAdjoint(int[][] mat) {
        int size = mat.length;
        int[][] adjoint = new int[size][size];
        if (size == 2) {
            adjoint[0][0] = mat[1][1];
            adjoint[1][1] = mat[0][0];
            adjoint[0][1] = -mat[0][1];
            adjoint[1][0] = -mat[1][0];
            return adjoint;
        }
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                int[][] minor = removeCopyMatrix(mat, i, j);
                adjoint[j][i] = ((i + j) % 2 == 0 ? 1 : -1) * findDeterminant(minor) % m;
                if (adjoint[j][i] < 0) adjoint[j][i] += m;
            }
        }
        return adjoint;
    }

    private int[][] scalarMultiply(int scalar, int[][] mat) {
        for (int i = 0; i < mat.length; i++) {
            for (int j = 0; j < mat[i].length; j++) {
                mat[i][j] = (scalar * mat[i][j]) % m;
                if (mat[i][j] < 0) mat[i][j] += m;
            }
        }
        return mat;
    }

    public String encrypt(String plaintext, int[][] keyMatrix) {
        int[][] plaintextMatrix = convertStrMat(plaintext, keyMatrix.length);
        System.out.println("Plaintext Matrix:");
        printMatrix(plaintextMatrix);
        int[][] cipherMatrix = findMatMul(keyMatrix, plaintextMatrix);
        System.out.println("Ciphertext Matrix:");
        printMatrix(cipherMatrix);
        return convertMatStr(cipherMatrix);
    }

    public String decrypt(String ciphertext, int[][] keyMatrix) {
        int det = findDeterminant(keyMatrix);
        if (det == 0 || findGCD(det, m) != 1) {
            throw new IllegalArgumentException("Matrix is not invertible");
        }
        int detInv = extendedEuclids(det, m, m);
        int[][] adjointMatrix = findAdjoint(keyMatrix);
        int[][] inverseMatrix = scalarMultiply(detInv, adjointMatrix);
        System.out.println("Inverse Key Matrix:");
        printMatrix(inverseMatrix);
        int[][] ciphertextMatrix = convertStrMat(ciphertext, inverseMatrix.length);
        int[][] plaintextMatrix = findMatMul(inverseMatrix, ciphertextMatrix);
        return convertMatStr(plaintextMatrix).replaceAll("X*$", "");
    }

    public int[][] knownPlaintextAttack(String pt, String ct, int size) {
        int[][] ptMatrix = convertStrMat(pt, size);
        System.out.println("Plaintext Matrix:");
        printMatrix(ptMatrix);

        int[][] ctMatrix = convertStrMat(ct, size);
        System.out.println("Ciphertext Matrix:");
        printMatrix(ctMatrix);

        int detCt = findDeterminant(ctMatrix);
        System.out.println("Determinant of Ciphertext Matrix: " + detCt);

        int detInv = extendedEuclids(detCt, m, m);
        System.out.println("Inverse of Determinant: " + detInv);

        int[][] adjCt = findAdjoint(ctMatrix);
        System.out.println("Adjoint of Ciphertext Matrix:");
        printMatrix(adjCt);

        int[][] ctInv = scalarMultiply(detInv, adjCt);
        System.out.println("Inverse of Ciphertext Matrix:");
        printMatrix(ctInv);

        int[][] keyInv = findMatMul(ptMatrix, ctInv);
        System.out.println("Inverse of Key Matrix:");
        printMatrix(keyInv);

        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the ciphertext to decrypt using the found key: ");
        String newCiphertext = scanner.nextLine().toUpperCase();
        int[][] newCiphertextMatrix = convertStrMat(newCiphertext, keyInv.length);
        int[][] newPlaintextMatrix = findMatMul(keyInv, newCiphertextMatrix);
        System.out.println("Decrypted plaintext Matrix using the found key:");
        printMatrix(newPlaintextMatrix);
        String PlainText = convertMatStr(newPlaintextMatrix);
        System.out.println("PlainText:" + PlainText);

        return keyInv;
    }

    private void printMatrix(int[][] matrix) {
        for (int[] row : matrix) {
            for (int elem : row) {
                System.out.print(elem + " ");
            }
            System.out.println();
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String alphaList = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        HC hillCipher = new HC(alphaList.length(), alphaList);

        while (true) {
            System.out.println("\nHill Cipher Menu");
            System.out.println("1. Encrypt");
            System.out.println("2. Decrypt");
            System.out.println("3. Known Plaintext Attack");
            System.out.println("4. Exit");
            System.out.print("Enter your choice: ");
            int choice = scanner.nextInt();
            scanner.nextLine(); // Consume newline

            switch (choice) {
                case 1 -> {
                    System.out.print("Enter the plaintext: ");
                    String plaintext = scanner.nextLine().toUpperCase();
                    System.out.print("Enter the size of the key matrix: ");
                    int size = scanner.nextInt();
                    int[][] keyMatrix = new int[size][size];
                    System.out.println("Enter the key matrix:");
                    for (int i = 0; i < size; i++) {
                        for (int j = 0; j < size; j++) {
                            keyMatrix[i][j] = scanner.nextInt();
                        }
                    }
                    String ciphertext = hillCipher.encrypt(plaintext, keyMatrix);
                    System.out.println("Ciphertext: " + ciphertext);
                }
                case 2 -> {
                    System.out.print("Enter the ciphertext: ");
                    String ciphertext = scanner.nextLine().toUpperCase();
                    System.out.print("Enter the size of the key matrix: ");
                    int size = scanner.nextInt();
                    int[][] keyMatrix = new int[size][size];
                    System.out.println("Enter the key matrix:");
                    for (int i = 0; i < size; i++) {
                        for (int j = 0; j < size; j++) {
                            keyMatrix[i][j] = scanner.nextInt();
                        }
                    }
                    String plaintext = hillCipher.decrypt(ciphertext, keyMatrix);
                    System.out.println("Decrypted Plaintext: " + plaintext);
                }
                case 3 -> {
                    System.out.print("Enter the known plaintext: ");
                    String pt = scanner.nextLine().toUpperCase();
                    System.out.print("Enter the corresponding ciphertext: ");
                    String ct = scanner.nextLine().toUpperCase();
                    System.out.print("Enter the size of the key matrix: ");
                    int size = scanner.nextInt();
                    hillCipher.knownPlaintextAttack(pt, ct, size);
                }
                case 4 -> {
                    System.out.println("Exiting...");
                    return;
                }
                default -> System.out.println("Invalid choice! Please select again.");
            }
        }
    }
}

File: Exp_2/hc.py
# Global variable for alphabet list
ALPHA_LIST = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
M = len(ALPHA_LIST)

def find_gcd(a, b):
    """Find the Greatest Common Divisor of two numbers."""
    if a < b:
        a, b = b, a
    while b != 0:
        a, b = b, a % b
    return a

def extended_euclids(a, b, m):
    """Implementation of Extended Euclidean Algorithm."""
    t1, t2 = 0, 1
    if a < b:
        a, b = b, a
    while b != 0:
        q = a // b
        a, b = b, a % b
        t1, t2 = t2, t1 - q * t2
    return t1 + m if t1 < 0 else t1

def convert_str_mat(word, syllable):
    """Convert string to matrix."""
    # Pad with 'X' if needed
    while len(word) % syllable != 0:
        word += 'X'
    
    cols = len(word) // syllable
    result = [[0] * cols for _ in range(syllable)]
    k = 0
    
    for i in range(cols):
        for j in range(syllable):
            result[j][i] = ALPHA_LIST.index(word[k])
            k += 1
    
    return result

def convert_mat_str(mat):
    """Convert matrix to string."""
    text = ""
    for i in range(len(mat[0])):
        for j in range(len(mat)):
            text += ALPHA_LIST[mat[j][i] % M]
    return text

def find_mat_mul(mat_a, mat_b):
    """Multiply two matrices."""
    rows_a = len(mat_a)
    cols_a = len(mat_a[0])
    cols_b = len(mat_b[0])
    result = [[0] * cols_b for _ in range(rows_a)]
    
    for i in range(rows_a):
        for j in range(cols_b):
            for k in range(cols_a):
                result[i][j] += mat_a[i][k] * mat_b[k][j]
            result[i][j] %= M
    
    return result

def remove_copy_matrix(mat, i, j):
    """Remove row i and column j from matrix."""
    return [[mat[row][col] for col in range(len(mat)) if col != j]
            for row in range(len(mat)) if row != i]

def check_square_matrix(mat):
    """Check if matrix is square."""
    return all(len(row) == len(mat) for row in mat)

def find_determinant(mat):
    """Calculate determinant of matrix."""
    if not check_square_matrix(mat):
        return -1
    
    if len(mat) == 2:
        return (mat[0][0] * mat[1][1] - mat[0][1] * mat[1][0]) % M
    
    det = 0
    for c in range(len(mat)):
        sign = 1 if c % 2 == 0 else -1
        det += sign * mat[0][c] * find_determinant(remove_copy_matrix(mat, 0, c))
        det %= M
    
    return det if det >= 0 else det + M

def find_adjoint(mat):
    """Find adjoint of matrix."""
    size = len(mat)
    adjoint = [[0] * size for _ in range(size)]
    
    if size == 2:
        adjoint[0][0] = mat[1][1]
        adjoint[1][1] = mat[0][0]
        adjoint[0][1] = -mat[0][1]
        adjoint[1][0] = -mat[1][0]
        return adjoint
    
    for i in range(size):
        for j in range(size):
            minor = remove_copy_matrix(mat, i, j)
            sign = 1 if (i + j) % 2 == 0 else -1
            adjoint[j][i] = (sign * find_determinant(minor)) % M
            if adjoint[j][i] < 0:
                adjoint[j][i] += M
    
    return adjoint

def scalar_multiply(scalar, mat):
    """Multiply matrix by scalar."""
    result = [[0] * len(mat[0]) for _ in range(len(mat))]
    for i in range(len(mat)):
        for j in range(len(mat[0])):
            result[i][j] = (scalar * mat[i][j]) % M
            if result[i][j] < 0:
                result[i][j] += M
    return result

def print_matrix(matrix):
    """Print matrix in readable format."""
    for row in matrix:
        print(" ".join(str(elem) for elem in row))

def encrypt(plaintext, key_matrix):
    """Encrypt plaintext using Hill cipher."""
    plaintext_matrix = convert_str_mat(plaintext, len(key_matrix))
    print("Plaintext Matrix:")
    print_matrix(plaintext_matrix)
    
    cipher_matrix = find_mat_mul(key_matrix, plaintext_matrix)
    print("Ciphertext Matrix:")
    print_matrix(cipher_matrix)
    
    return convert_mat_str(cipher_matrix)

def decrypt(ciphertext, key_matrix):
    """Decrypt ciphertext using Hill cipher."""
    det = find_determinant(key_matrix)
    if det == 0 or find_gcd(det, M) != 1:
        raise ValueError("Matrix is not invertible")
    
    det_inv = extended_euclids(det, M, M)
    adjoint_matrix = find_adjoint(key_matrix)
    inverse_matrix = scalar_multiply(det_inv, adjoint_matrix)
    
    print("Inverse Key Matrix:")
    print_matrix(inverse_matrix)
    
    ciphertext_matrix = convert_str_mat(ciphertext, len(inverse_matrix))
    plaintext_matrix = find_mat_mul(inverse_matrix, ciphertext_matrix)
    
    return convert_mat_str(plaintext_matrix).rstrip('X')

def known_plaintext_attack(pt, ct, size):
    """Perform known plaintext attack."""
    pt_matrix = convert_str_mat(pt, size)
    print("Plaintext Matrix:")
    print_matrix(pt_matrix)

    ct_matrix = convert_str_mat(ct, size)
    print("Ciphertext Matrix:")
    print_matrix(ct_matrix)

    det_ct = find_determinant(ct_matrix)
    print(f"Determinant of Ciphertext Matrix: {det_ct}")

    det_inv = extended_euclids(det_ct, M, M)
    print(f"Inverse of Determinant: {det_inv}")

    adj_ct = find_adjoint(ct_matrix)
    print("Adjoint of Ciphertext Matrix:")
    print_matrix(adj_ct)

    ct_inv = scalar_multiply(det_inv, adj_ct)
    print("Inverse of Ciphertext Matrix:")
    print_matrix(ct_inv)

    key_inv = find_mat_mul(pt_matrix, ct_inv)
    print("Inverse of Key Matrix:")
    print_matrix(key_inv)

    new_ciphertext = input("Enter the ciphertext to decrypt using the found key: ").upper()
    new_ciphertext_matrix = convert_str_mat(new_ciphertext, len(key_inv))
    new_plaintext_matrix = find_mat_mul(key_inv, new_ciphertext_matrix)
    print("Decrypted plaintext Matrix using the found key:")
    print_matrix(new_plaintext_matrix)
    plaintext = convert_mat_str(new_plaintext_matrix)
    print(f"PlainText: {plaintext}")

    return key_inv

def main():
    """Main function with menu interface."""
    while True:
        print("\nHill Cipher Menu")
        print("1. Encrypt")
        print("2. Decrypt")
        print("3. Known Plaintext Attack")
        print("4. Exit")
        
        try:
            choice = int(input("Enter your choice: "))
            
            if choice == 1:
                plaintext = input("Enter the plaintext: ").upper()
                size = int(input("Enter the size of the key matrix: "))
                print("Enter the key matrix:")
                key_matrix = [[int(input(f"Enter element [{i}][{j}]: ")) 
                             for j in range(size)] for i in range(size)]
                ciphertext = encrypt(plaintext, key_matrix)
                print(f"Ciphertext: {ciphertext}")
                
            elif choice == 2:
                ciphertext = input("Enter the ciphertext: ").upper()
                size = int(input("Enter the size of the key matrix: "))
                print("Enter the key matrix:")
                key_matrix = [[int(input(f"Enter element [{i}][{j}]: ")) 
                             for j in range(size)] for i in range(size)]
                plaintext = decrypt(ciphertext, key_matrix)
                print(f"Decrypted Plaintext: {plaintext}")
                
            elif choice == 3:
                pt = input("Enter the known plaintext: ").upper()
                ct = input("Enter the corresponding ciphertext: ").upper()
                size = int(input("Enter the size of the key matrix: "))
                known_plaintext_attack(pt, ct, size)
                
            elif choice == 4:
                print("Exiting...")
                break
                
            else:
                print("Invalid choice! Please select again.")
                
        except ValueError as e:
            print(f"Error: {e}")
            print("Please try again with valid input.")

if __name__ == "__main__":
    main()